alacritty/src/display.rs

```diff
@@ -0,0 +1,420 @@
+// Copyright 2016 Joe Wilm, The Alacritty Project Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The display subsystem including window management, font rasterization, and
+//! GPU drawing.
+use std::f64;
+
+use glutin::dpi::{PhysicalPosition, PhysicalSize};
+use glutin::{ContextCurrentState, NotCurrent, PossiblyCurrent, RawContext};
+use log::info;
+
+use font::{self, Rasterize};
+
+use alacritty_terminal::config::{Font, StartupMode};
+use alacritty_terminal::index::Line;
+use alacritty_terminal::message_bar::Message;
+use alacritty_terminal::meter::Meter;
+use alacritty_terminal::renderer::rects::{RenderLines, RenderRect};
+use alacritty_terminal::renderer::{self, GlyphCache, QuadRenderer};
+use alacritty_terminal::term::color::Rgb;
+use alacritty_terminal::term::{RenderableCell, SizeInfo};
+
+use crate::config::Config;
+use crate::event::EventProxy;
+use crate::window::{self, Window};
+
+#[derive(Debug)]
+pub enum Error {
+    /// Error with window management
+    Window(window::Error),
+
+    /// Error dealing with fonts
+    Font(font::Error),
+
+    /// Error in renderer
+    Render(renderer::Error),
+
+    /// Error during buffer swap
+    ContextError(glutin::ContextError),
+}
+
+impl ::std::error::Error for Error {
+    fn cause(&self) -> Option<&dyn (::std::error::Error)> {
+        match *self {
+            Error::Window(ref err) => Some(err),
+            Error::Font(ref err) => Some(err),
+            Error::Render(ref err) => Some(err),
+            Error::ContextError(ref err) => Some(err),
+        }
+    }
+
+    fn description(&self) -> &str {
+        match *self {
+            Error::Window(ref err) => err.description(),
+            Error::Font(ref err) => err.description(),
+            Error::Render(ref err) => err.description(),
+            Error::ContextError(ref err) => err.description(),
+        }
+    }
+}
+
+impl ::std::fmt::Display for Error {
```

Comment:
I think we should move away from this syntax. We're already writing `fmt::Display` and stuff like that in other places. What do you think?

Reply:
Completely agree. Especially the `::std` syntax should be removed, though this code is just copied not rewritten.

I've just gone ahead and fixed it anyways though.